diff -rupN ../../grlib/grlib-gpl-1.3.7-b4144/designs/leon3-xilinx-ml605/Makefile work/Makefile
--- ../../grlib/grlib-gpl-1.3.7-b4144/designs/leon3-xilinx-ml605/Makefile	2014-04-16 16:50:08.000000000 +0200
+++ work/Makefile	2015-01-04 02:21:07.651044424 +0100
@@ -1,5 +1,5 @@
 include .config
-GRLIB=../..
+GRLIB=../../../grlib/grlib-gpl-1.3.7-b4144
 TOP=leon3mp
 DESIGN=leon3-xilinx-ml605
 BOARD=xilinx-ml605-xc6vlx240t
@@ -14,6 +14,7 @@ EFFORT=high
 XSTOPT=-uc leon3mp.xcf
 SYNPOPT="set_option -pipe 1; set_option -retiming 0; set_option -enable_prepacking 1;\
 set_option -write_apr_constraint 0;set_option -resource_sharing 1; set_option -symbolic_fsm_compiler 1"
+
 VHDLOPTSYNFILES= \
 	mig_37/user_design/rtl/controller/arb_mux.vhd \
 	mig_37/user_design/rtl/controller/arb_row_col.vhd \
@@ -69,58 +70,6 @@ VHDLOPTSYNFILES= \
 	pcie/v6_pcie_v1_7/source/*.vhd \
 	$(GRLIB)/lib/gaisler/pcie/pcie.vhd
 
-ifeq ("$(CONFIG_PCIEXP_MASTER_TARGET)","y")
-  ifeq ("$(CONFIG_LANE_WIDTH1)","y")
-  UCF+=pcie_ucf/pcie_master_target_lane1.ucf
-  UCF_PLANAHEAD+=pcie_ucf/pcie_master_target_lane1.ucf
-  endif
-  ifeq ("$(CONFIG_LANE_WIDTH2)","y")
-  UCF+=pcie_ucf/pcie_master_target_lane2.ucf
-  UCF_PLANAHEAD+=pcie_ucf/pcie_master_target_lane2.ucf
-  endif
-  ifeq ("$(CONFIG_LANE_WIDTH4)","y")
-  UCF+=pcie_ucf/pcie_master_target_lane4.ucf
-  UCF_PLANAHEAD+=pcie_ucf/pcie_master_target_lane4.ucf
-  endif
-  ifeq ("$(CONFIG_LANE_WIDTH8)","y")
-  UCF+=pcie_ucf/pcie_master_target_lane8.ucf
-  UCF_PLANAHEAD+=pcie_ucf/pcie_master_target_lane8.ucf
-  endif
-VHDLOPTSYNFILES+=$(GRLIB)/lib/gaisler/pcie/pcie_master_target_virtex.vhd
-endif
-ifeq ("$(CONFIG_PCIEXP_MASTER_FIFO)","y")
-  ifeq ("$(CONFIG_PCIEXP_MASTER_FIFO_DMA)","y")
-    ifeq ("$(CONFIG_LANE_WIDTH1)","y")
-    UCF+=pcie_ucf/pcie_master_fifo_dma_lane1.ucf
-    UCF_PLANAHEAD+=pcie_ucf/pcie_master_fifo_dma_lane1.ucf
-    endif
-    ifeq ("$(CONFIG_LANE_WIDTH2)","y")
-    UCF+=pcie_ucf/pcie_master_fifo_dma_lane2.ucf
-    UCF_PLANAHEAD+=pcie_ucf/pcie_master_fifo_dma_lane2.ucf
-    endif
-    ifeq ("$(CONFIG_LANE_WIDTH4)","y")
-    UCF+=pcie_ucf/pcie_master_fifo_dma_lane4.ucf
-    UCF_PLANAHEAD+=pcie_ucf/pcie_master_fifo_dma_lane4.ucf
-    endif
-VHDLOPTSYNFILES+=$(GRLIB)/lib/gaisler/pcie/pcie_master_fifo_virtex.vhd
-VHDLOPTSYNFILES+=$(GRLIB)/lib/gaisler/pcie/pciedma.vhd
-  endif
-  ifneq ("$(CONFIG_PCIEXP_MASTER_FIFO_DMA)","y")
-    ifeq ("$(CONFIG_LANE_WIDTH1)","y")
-    UCF+=pcie_ucf/pcie_master_fifo_lane1.ucf
-    UCF_PLANAHEAD+=pcie_ucf/pcie_master_fifo_lane1.ucf
-    endif
-    ifeq ("$(CONFIG_LANE_WIDTH2)","y")
-    UCF+=pcie_ucf/pcie_master_fifo_lane2.ucf
-    UCF_PLANAHEAD+=pcie_ucf/pcie_master_fifo_lane2.ucf
-    endif
-    ifeq ("$(CONFIG_LANE_WIDTH4)","y")
-    UCF+=pcie_ucf/pcie_master_fifo_lane4.ucf
-    UCF_PLANAHEAD+=pcie_ucf/pcie_master_fifo_lane4.ucf
-    endif
-VHDLOPTSYNFILES+=$(GRLIB)/lib/gaisler/pcie/pcie_master_fifo_virtex.vhd
-  endif
-endif
 
 VERILOGOPTSYNFILES= \
 	mig/user_design/rtl/controller/*.v \
@@ -136,23 +85,12 @@ VHDLSYNFILES= \
 
 VHDLSIMFILES=testbench.vhd
 
-ifeq ("$(CONFIG_PCIEXP_MASTER_TARGET)","y")
-VHDLSIMFILES+=$(GRLIB)/lib/tech/unisim/ise/GTPA1_DUAL.vhd $(GRLIB)/lib/tech/unisim/ise/GTP_DUAL.vhd \
-$(GRLIB)/lib/tech/unisim/ise/GTXE1.vhd $(GRLIB)/lib/tech/unisim/ise/GTX_DUAL.vhd \
-$(GRLIB)/lib/tech/unisim/ise/PCIE_2_0.vhd $(GRLIB)/lib/tech/unisim/ise/PCIE_A1.vhd
-else 
-ifeq ("$(CONFIG_PCIEXP_MASTER_FIFO)","y")
-VHDLSIMFILES+=$(GRLIB)/lib/tech/unisim/ise/GTPA1_DUAL.vhd $(GRLIB)/lib/tech/unisim/ise/GTP_DUAL.vhd \
-$(GRLIB)/lib/tech/unisim/ise/GTXE1.vhd $(GRLIB)/lib/tech/unisim/ise/GTX_DUAL.vhd \
-$(GRLIB)/lib/tech/unisim/ise/PCIE_2_0.vhd $(GRLIB)/lib/tech/unisim/ise/PCIE_A1.vhd
-endif
-endif
-
 SIMTOP=testbench
 SDCFILE=default.sdc
 BITGEN=$(GRLIB)/boards/$(BOARD)/default.ut
 CLEAN=soft-clean migclean pcieclean
 TECHLIBS = secureip unisim
+EXTRALIBS=../../../lib
 VLOGOPT= -O0
 VCOMOPT= -explicit -O0 
 ifeq ("$(GRLIB_SIMULATOR)","ALDEC")
