diff --git a/include/configs/zynq_common.h b/include/configs/zynq_common.h
index 971e176..765242f 100644
--- a/include/configs/zynq_common.h
+++ b/include/configs/zynq_common.h
@@ -182,9 +182,9 @@
 /* Default environment */
 #define CONFIG_EXTRA_ENV_SETTINGS	\
 	"ethaddr=00:0a:35:00:01:22\0"	\
-	"kernel_image=uImage\0"	\
+	"kernel_image=zImage\0"	\
 	"ramdisk_image=uramdisk.image.gz\0"	\
-	"devicetree_image=devicetree.dtb\0"	\
+	"devicetree_image=devicetree_linaro.dtb\0"	\
 	"kernel_size=0x500000\0"	\
 	"devicetree_size=0x20000\0"	\
 	"ramdisk_size=0x5E0000\0"	\
@@ -207,8 +207,7 @@
 		"mmcinfo;" \
 		"fatload mmc 0 0x3000000 ${kernel_image};" \
 		"fatload mmc 0 0x2A00000 ${devicetree_image};" \
-		"fatload mmc 0 0x2000000 ${ramdisk_image};" \
-		"bootm 0x3000000 0x2000000 0x2A00000\0" \
+		"bootm 0x3000000 - 0x2A00000\0" \
 	"nandboot=echo Copying Linux from NAND flash to RAM...;" \
 		"nand read 0x3000000 0x100000 ${kernel_size};" \
 		"nand read 0x2A00000 0x600000 ${devicetree_size};" \
@@ -222,7 +221,7 @@
 		"bootm 0x3000000 0x2000000 0x2A00000\0"
 
 /* default boot is according to the bootmode switch settings */
-#define CONFIG_BOOTCOMMAND "run modeboot"
+#define CONFIG_BOOTCOMMAND "run sdboot"
 
 #define CONFIG_BOOTDELAY	3 /* -1 to Disable autoboot */
 
